add VAD task and VAD recipe for AMI

Some-random opened the pull request.

sw005320 commented:
@ftshijt, can you review this PR?

ftshijt commented on egs2/TEMPLATE/vad1/vad.sh:
```diff
@@ -0,0 +1,947 @@
+
+        if [ -n "${post_process_local_data_opts}" ]; then
+            # Do any additional local data post-processing here
+            local/data.sh ${post_process_local_data_opts} --asr_data_dir "${data_feats}/${train_set}"
```
What kind of post-process we would like to do here? I did not find related functions in local/data.sh. I personally. think it would be better to get it generalized and put into pyscripts or scripts for consistency

ftshijt replied:
Could you help checking whether we can integrate it in to local/data.sh or it would be better to be designed as a general scripts?

ftshijt commented on egs2/TEMPLATE/vad1/vad.sh:
```diff
@@ -0,0 +1,947 @@
+            _opts+="--train_data_path_and_name_and_type ${_split_dir}/${_scp},speech,${_type} "
+            _opts+="--train_shape_file ${_split_dir}/speech_shape "
+            # shellcheck disable=SC2068
+            for i in ${!ref_text_names[@]}; do
+                _opts+="--fold_length ${vad_text_fold_length} "
+                _opts+="--train_data_path_and_name_and_type ${_split_dir}/${ref_text_files[$i]},${ref_text_names[$i]},text "
+                _opts+="--train_shape_file ${_split_dir}/${ref_text_names[$i]}_shape "
+            done
+            _opts+="--multiple_iterator true "
```
Please double check if this is needed.

ftshijt commented on egs2/TEMPLATE/vad1/vad.sh:
```diff
@@ -0,0 +1,947 @@
+                    >"${data_feats}/${dset}/feats.scp"
+            fi
+
+            # Remove empty text
+            # shellcheck disable=SC2068
+            for ref_txt in ${ref_text_files[@]}; do
+                <"${data_feats}/org/${dset}/${ref_txt}" \
+                    awk ' { if( NF != 1 ) print $0; } ' >"${data_feats}/${dset}/${ref_txt}"
+            done
```
Please double check if this is needed.

ftshijt commented on egs2/TEMPLATE/vad1/vad.sh:
```diff
@@ -0,0 +1,947 @@
+                fi
+                utils/copy_data_dir.sh --validate_opts --non-print data/"${dset}" "${data_feats}${_suf}/${dset}"
+
+                # Copy reference text files if there is more than 1 reference
+                # shellcheck disable=SC2068
+                if [ ${#ref_text_files[@]} -gt 1 ]; then
+                    for ref_txt in ${ref_text_files[@]}; do
+                        [ -f data/${dset}/${ref_txt} ] && cp data/${dset}/${ref_txt} ${data_feats}${_suf}/${dset}
+                    done
+                fi
```
Please double check if this is needed.

Some-random replied:
Deleted

ftshijt commented on egs2/TEMPLATE/vad1/vad.sh:
```diff
@@ -0,0 +1,947 @@
+                # 1. Copy datadir
+                utils/copy_data_dir.sh --validate_opts --non-print data/"${dset}" "${data_feats}${_suf}/${dset}"
+
+                # Copy reference text files if there is more than 1 reference
+                if [ ${#ref_text_files[@]} -gt 1 ]; then
+                    # shellcheck disable=SC2068
+                    for ref_txt in ${ref_text_files[@]}; do
+                        [ -f data/${dset}/${ref_txt} ] && cp data/${dset}/${ref_txt} ${data_feats}${_suf}/${dset}
+                    done
+                fi
```
Please double check if this is needed.

Some-random replied:
Deleted

ftshijt commented on egs2/TEMPLATE/vad1/vad.sh:
```diff
@@ -0,0 +1,947 @@
+
+        _opts+="--train_data_path_and_name_and_type ${_vad_train_dir}/${_scp},speech,${_type} "
+        _opts+="--valid_data_path_and_name_and_type ${_vad_valid_dir}/${_scp},speech,${_type} "
+        # shellcheck disable=SC2068
+        for i in ${!ref_text_files[@]}; do
+            _opts+="--train_data_path_and_name_and_type ${_vad_train_dir}/${ref_text_files[$i]},${ref_text_names[$i]},text "
+            _opts+="--valid_data_path_and_name_and_type ${_vad_valid_dir}/${ref_text_files[$i]},${ref_text_names[$i]},text "
+        done
+
```
Please double check if this is needed.

ftshijt commented on egs2/TEMPLATE/vad1/vad.sh:
```diff
@@ -0,0 +1,947 @@
+                     _opts+="--train_data_path_and_name_and_type ${_vad_train_dir}/${aux_dset},text,text "
+                done
+            fi
+            # shellcheck disable=SC2068
+            for i in ${!ref_text_names[@]}; do
+                _opts+="--fold_length ${vad_text_fold_length} "
+                _opts+="--train_data_path_and_name_and_type ${_vad_train_dir}/${ref_text_files[$i]},${ref_text_names[$i]},text "
+                _opts+="--train_shape_file ${vad_stats_dir}/train/${ref_text_names[$i]}_shape "
+            done
+        fi
+
+        # shellcheck disable=SC2068
+        for i in ${!ref_text_names[@]}; do
+            _opts+="--valid_data_path_and_name_and_type ${_vad_valid_dir}/${ref_text_files[$i]},${ref_text_names[$i]},text "
+            _opts+="--valid_shape_file ${vad_stats_dir}/valid/${ref_text_names[$i]}_shape "
+        done
```
Please double check if this is needed.

ftshijt commented on egs2/TEMPLATE/vad1/vad.sh:
```diff
@@ -0,0 +1,947 @@
+
+            done
+
+            ${python} -m espnet2.bin.vad_scoring \
```
Usually, for scoring, we would prefer putting them to pyscripts or scripts instead of binary

ftshijt commented on espnet2/train/preprocessor.py:
```diff
@@ -442,6 +442,55 @@ def _text_process(
+            # transform float time intervals to integers of length len(speech)
+            text = data[self.text_name]
+            text_split = [float(item) * 100 for item in text.split(" ")]
+            # input is split to 10s chunks, multiply the original timestamp by 100
+            # text_array = np.zeros([1000, 2], dtype=np.int64)
+            # if len(text_split) != 0:
+            #     for i in range(0, len(text_split), 2):
+            #         text_array[int(text_split[i]):int(text_split[i+1]), :] = 1
+            #     text_array[:, 1] = 1 - text_array[:, 0]
+            text_array = np.zeros([1001], dtype=np.int64)
```
Could you remove the hard code settings?

ftshijt commented on egs2/ami/vad1/local/generate_vad_file.py:
```diff
@@ -0,0 +1,90 @@
+        duration = wav_file.getnframes() / float(wav_file.getframerate())
+    # split the whole wave file into 10s segments
+    for i in range(int(duration / 10)):
+        start = i * 10
+        end = (i + 1) * 10
```
Why not use chunk iterator for chunking, we could have more options then (e.g., random chunk size etc)

ftshijt reviewed (changes requested):
Thanks for the contribution. Some comments as follow

Some-random commented on egs2/TEMPLATE/vad1/vad.sh:
```diff
@@ -0,0 +1,947 @@
+audio_format=flac    # Audio format: wav, flac, wav.ark, flac.ark  (only in feats_type=raw).
+fs=16k               # Sampling rate.
+min_wav_duration=0.1 # Minimum duration in second.
+max_wav_duration=20  # Maximum duration in second.
```
This is the length of chunked speech segments, current the length is 10s, pyannote uses 5s, so I think it should be enough

Some-random commented on egs2/TEMPLATE/vad1/vad.sh:
```diff
@@ -0,0 +1,947 @@
+
+    # Data preparation related
+    --local_data_opts # The options given to local/data.sh (default="${local_data_opts}").
+
```
The new arguments I added in run.sh are specific to AMI data, it mainly specifies the different stages in AMI data extraction. As different datasets will have different local_data_opts, I don't think additional help message is needed

Some-random commented on egs2/TEMPLATE/vad1/vad.sh:
```diff
@@ -0,0 +1,947 @@
+    ref_text_files_str="text "
+    ref_text_names_str="text "
+else
+    # For multiple speakers, text file path and name are text_spk[1-N] and [text, text_spk2, ...]
+    #TODO(simpleoier): later to support flexibly defined text prefix
+    ref_text_files_str="text_spk1 "
+    ref_text_names_str="text "
+    for n in $(seq 2 ${num_ref}); do
+        ref_text_files_str+="text_spk${n} "
+        ref_text_names_str+="text_spk${n} "
+    done
```
You're right, we don't consider multiple spk.

Some-random commented on egs2/TEMPLATE/vad1/vad.sh:
```diff
@@ -0,0 +1,947 @@
+                rm -f ${data_feats}${_suf}/${dset}/{segments,wav.scp,reco2file_and_channel,reco2dur}
+
+                # Copy reference text files if there is more than 1 reference
+                if [ ${#ref_text_files[@]} -gt 1 ]; then
+                    # shellcheck disable=SC2068
+                    for ref_txt in ${ref_text_files[@]}; do
+                        [ -f data/${dset}/${ref_txt} ] && cp data/${dset}/${ref_txt} ${data_feats}${_suf}/${dset}
+                    done
+                fi
```
Sure

Some-random commented:
> What is the current status, @Some-random?

Hi Shinji, It appears that @ftshijt and I have reached a consensus on the majority of corrections needed for my code. However, there are still two unresolved points where I'm awaiting further guidance:

1  Would it be acceptable to place the scoring script in "espnet2.bin.vad_scoring"?
2  Can we forgo the use of a chunk iterator for this PR? Given the substantial changes introduced in this PR, I suggest that we initially proceed with a fixed segment. The adjustment to the chunk iterator can be implemented in a later iteration.

Once I receive additional instructions and address these issues, I plan to perform an end-to-end check. This will ensure that everything is running smoothly and is in harmony with the latest code updates.

sw005320 commented:
> 1 Would it be acceptable to place the scoring script in "espnet2.bin.vad_scoring"? 

OK, since SE is also doing the same thing.

> 2 Can we forgo the use of a chunk iterator for this PR? Given the substantial changes introduced in this PR, I suggest that we initially proceed with a fixed segment. The adjustment to the chunk iterator can be implemented in a later iteration.

OK

Some-random commented:
Hi @sw005320 , sorry for the delay. I have successfully executed the entire recipe and resolved a few bugs in the process. It should be ready for merging at this point. Please don't hesitate to let me know if there are any additional issues.

sw005320 commented:
@ftshijt, can you review this PR again?

@Some-random
There are some errors.
https://github.com/espnet/espnet/actions/runs/5221471893/jobs/9425771502?pr=5156#step:18:103678
I recommend you check it locally in advance to avoid this type of error, especially for 
https://github.com/espnet/espnet/blob/master/ci/test_integration_espnet2.sh#L302-L309

sw005320 commented:
Please fix https://github.com/espnet/espnet/actions/runs/5230271708/jobs/9443727951?pr=5156#step:12:103899

ftshijt commented:
> Given the substantial changes introduced in this PR, I suggest that we initially proceed with a fixed segment. The adjustment to the chunk iterator can be implemented in a later iteration.

In the case if you would like to change it in the future and it is on the plan, it would be great if you simply add a `TODO` comment so that we can get to know your next movement.

ftshijt commented on egs2/TEMPLATE/vad1/vad.sh:
```diff
@@ -0,0 +1,905 @@
+    log "Skip the uploading stage"
+fi
+
+if ! "${skip_upload_hf}"; then
```
Have you tested this out?

ftshijt commented on egs2/ami/vad1/conf/tuning/decode_rnn.yaml:
```diff
@@ -0,0 +1,2 @@
+maxlenratio: 0.0
+minlenratio: 0.0
```
Do we need to control maxlen and minlen for the vad case? Since it is not an auto-regressive generation protocal, might be not?

ftshijt commented on egs2/ami/vad1/conf/tuning/decode_transformer.yaml:
```diff
@@ -0,0 +1,4 @@
+batch_size: 1
+beam_size: 10
+ctc_weight: 0.2
+lm_weight: 0.2
```
Can you double check if these arguments are used?

ftshijt commented on egs2/ami/vad1/conf/tuning/train_vad_rnn.yaml:
```diff
@@ -0,0 +1,56 @@
+    subsample: [1, 1, 1, 1]
+
+model_conf:
+    length_normalized_loss: false
```
do we need to use length normalized loss in the context?

ftshijt commented on egs2/ami/vad1/run.sh:
```diff
@@ -0,0 +1,35 @@
+    --train_set "${train_set}" \
+    --valid_set "${valid_set}" \
+    --test_sets "${test_sets}" \
+    --local_data_opts "--stage 2" \
+    --gpu_inference true
```
Would be great if you can make them as the default (and let user to change them depends on their needs)

ftshijt commented on egs2/ami/vad1/scripts:
```diff
@@ -0,0 +1 @@
+../asr1/scripts
```
This is OK, but why not link to the template one (as the setup.sh should automatically link them?)

ftshijt commented on egs2/ami/vad1/steps:
```diff
@@ -0,0 +1 @@
+../asr1/steps
```
Ditto for others

ftshijt commented on espnet2/bin/vad_inference.py:
```diff
@@ -0,0 +1,469 @@
+        vad_train_config: Union[Path, str] = None,
+        vad_model_file: Union[Path, str] = None,
+        device: str = "cpu",
+        maxlenratio: float = 0.0,
+        minlenratio: float = 0.0,
```
Since we do not do beam search, I don't think these arguments are necessary

ftshijt commented on espnet2/bin/vad_inference.py:
```diff
@@ -0,0 +1,469 @@
+        self,
+        enc: np.ndarray,
+        threshold: float,
+        silence_to_speech_thresh: int,
+        speech_to_silence_thresh: int,
```
Just curious, do you have some references for using the two thresholds.

ftshijt commented on espnet2/bin/vad_inference.py:
```diff
@@ -0,0 +1,469 @@
+
+            # Only supporting batch_size==1
+            key = keys[0]
+            print(key)
```
```suggestion
```

ftshijt commented on espnet2/bin/vad_scoring.py:
```diff
@@ -0,0 +1,102 @@
+        hyp_array = [0 for _ in range(1000)]
+
+        ref_time_stamps = [int(float(item) * 100) for item in time_stamp_ref.split(" ")]
+        for j in range(0, len(ref_time_stamps), 2):
+            for k in range(ref_time_stamps[j], ref_time_stamps[j + 1]):
+                ref_array[k] = 1
```
Can we directly use array instead of iteration?

ftshijt commented on espnet2/bin/vad_scoring.py:
```diff
@@ -0,0 +1,102 @@
+            print(hyp_time_stamps)
+            continue
+
+        for j in range(0, len(hyp_time_stamps), 2):
+            for k in range(hyp_time_stamps[j], hyp_time_stamps[j + 1]):
+                hyp_array[k] = 1
+
+        # calculate the P and R
+        TP, FP, FN = 0, 0, 0
+        for j in range(1000):
+            if ref_array[j] == 1 and hyp_array[j] == 1:
+                TP += 1
+            elif ref_array[j] == 0 and hyp_array[j] == 1:
+                FP += 1
+            elif ref_array[j] == 1 and hyp_array[j] == 0:
+                FN += 1
```
Ditto

ftshijt commented on espnet2/tasks/vad.py:
```diff
@@ -0,0 +1,329 @@
+    default="utterance_mvn",
+    optional=True,
+)
+model_choices = ClassChoices(
+    "model",
+    classes=dict(
+        espnet=ESPnetVADModel,
+    ),
+    type_check=AbsESPnetModel,
+    default="espnet",
+)
```
Since we only have one choice here, I would suggest simply use the way in diar without `model_choices`

ftshijt commented on espnet2/vad/espnet_model.py:
```diff
@@ -0,0 +1,196 @@
+
+
+class ESPnetVADModel(ESPnetASRModel):
+    """RNN frame-level model"""
```
It does not limit to RNN right?

ftshijt commented on espnet2/vad/espnet_model.py:
```diff
@@ -0,0 +1,196 @@
+        encoder_out = torch.argmax(encoder_out, dim=1)
+        numerator = torch.sum(encoder_out == ys_out)
+        denominator = ys_out.numel()
+        acc = float(numerator) / float(denominator)
```
As the label would be bias, accuracy might not be a great indicator in some cases, how about adding more stats in each steps to help with the testing and training.

ftshijt reviewed (changes requested):
Thanks for the update! I have my review comments as follows:

Some-random commented:
Hello @sw005320 and @ftshijt , I've addressed the issues you pointed out and the updated code now passes the continuous integration checks. When you have a moment, could you kindly take another look? Thanks.

sw005320 commented on egs2/TEMPLATE/vad1/vad.sh:
There are some comments about `lm`.
You should modify them.

sw005320 commented on egs2/ami/vad1/README.md:
- Add a HF model link
- The table looks strange. Please fix it.
<img width="992" alt="image" src="https://github.com/espnet/espnet/assets/11741550/6ff9adf0-0b65-4c6f-b0b3-2e7b91287c31">

sw005320 commented on espnet2/bin/vad_inference.py:
```diff
@@ -0,0 +1,443 @@
+                Currently, the tags of espnet_model_zoo are supported.
+
+        Returns:
+            Speech2Text: Speech2Text instance.
```
```suggestion
            VoiceActivityDetect: VoiceActivityDetect instance.
```

sw005320 replied:
Please carefully check this file and modify similar mistakes accordingly.

sw005320 commented on espnet2/bin/vad_inference.py:
```diff
@@ -0,0 +1,443 @@
+        "--silence_to_speech_thresh",
+        type=int,
+        default=3,
+        help="Threshold for silence to speech",
```
Add more explanations.

sw005320 commented on espnet2/bin/vad_inference.py:
```diff
@@ -0,0 +1,443 @@
+        "--speech_to_silence_thresh",
+        type=int,
+        default=15,
+        help="Threshold for speech to silence",
```
Add more explanations.

sw005320 commented on espnet2/bin/vad_scoring.py:
Do you have a test script for this?
Please add them and also check with several examples.

Evaluation metrics must NOT have any mistakes.

sw005320 commented on espnet2/train/preprocessor.py:
Many parts in this file do not seem to go through the test.
Can you increase the test coverage?

sw005320 reviewed (commented):
Please add unit tests and integration tests.
Also, please add some explanations about the data format (especially for the output).

sw005320 commented:
@Some-random, it’s almost there.
So, can you reflect the comments?

sw005320 commented:
@Some-random, this is a reminder.

sw005320 commented:
@ftshijt, could I ask you to review this PR again?
I think this becomes a good shape now.

sw005320 commented:
@ftshijt, could you review this PR, again?

ftshijt commented:
Sorry for the lateness. I double checked the PR and it looks great to go. Let's merge it after CI passing.

ftshijt reviewed (approved):
I feel it is ready to merge.
